Central workflow to trigger pre-release workflows

prudhvigodithi opened the pull request:
### Description
Part of [pre-release](https://github.com/opensearch-project/opensearch-build/blob/main/RELEASE_PROCESS_OPENSEARCH.md#pre-release) actives, a central workflow that triggers all the pre-release workflows.

### Issues Resolved
https://github.com/opensearch-project/opensearch-build/issues/4136#issue-1948209066

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

prudhvigodithi commented:
- Basic approach to trigger all the jobs in order.
- Uses `wait: true`, waits until the job is completed, we can change to `false` once we are confident to ensure not to wait for job status.
- Can be improved in 2nd iteration to optimize the code to run `distribution-promote-artifact`  based out of lists like [OS, OSD], [tar, deb, rpm, windows, apt, yum].

prudhvigodithi commented:
FYI, Found that the Jenkins tests has a bug handling `build step` which  returns `parameters=[null, null, null, null]`, but the actual logic to invoke the build works with right parameters passed.

Seen this in past PR as well.
https://github.com/opensearch-project/opensearch-build-libraries/commit/35267f6b60a205947d125453d36dd75f26539ff0#diff-2691a2da3434bcea33ddddd273112adeddd1b30dfa84b8837ca556a509a0b668

I have raised an issue in JenkinsPipelineUnit repo. https://github.com/jenkinsci/JenkinsPipelineUnit/issues/551#issuecomment-1867018101

Divyaasm commented on jenkins/promotion/release-promotion.jenkinsfile:
```diff
@@ -0,0 +1,433 @@
+                    image 'docker/library/alpine:3'
+                    registryUrl 'https://public.ecr.aws/'
+                    alwaysPull true
+                }
```
LGTM overall.  Since the docker agent is same for all the triggered runs I would recommend we loop through the lists [OS, OSD], [tar, deb, rpm, windows, apt, yum] w.r.t. [x64, arm64] for all the possible mappings instead of individually passing parameters for each run, as part of optimization in 2nd iteration.

prudhvigodithi replied:
Those combinations are never unique @Divyaasm, also If I loop do you think I can still visualize as individual stage ? My idea was each stage should an individual promotion job, would be easy debug/verify if a promotion job is failed and run it manually as we do today.

Divyaasm replied:
I see, currently we have implemented this parallelly in build and validation workflows. Need to figure out ways to do that if we opt to loop the possible mappings.

gaiksaya commented on jenkins/promotion/release-promotion.jenkinsfile:
```diff
@@ -0,0 +1,433 @@
+            trim: true
+        )
+        string(
+            name: 'OPENSEARCH_RC',
+            description: 'OpenSearch Release Candidate',
```
```suggestion
            name: 'OPENSEARCH_RC_BUILD_NUMBER',
            description: 'OpenSearch Release Candidate Build Number',
```

gaiksaya commented on jenkins/promotion/release-promotion.jenkinsfile:
```diff
@@ -0,0 +1,433 @@
+            trim: true
+        )
+        string(
+            name: 'OPENSEARCH_DASHBOARDS_RC',
+            description: 'OpenSearch Dashboards Release Candidate',
```
```suggestion
            name: 'OPENSEARCH_DASHBOARDS_RC_BUILD_NUMBER',
            description: 'OpenSearch Dashboards Release Candidate Build Number',
```

gaiksaya commented on jenkins/promotion/release-promotion.jenkinsfile:
```diff
@@ -0,0 +1,433 @@
+                }
+            }
+        }
+        stage('OpenSearch Apt promotion') {
```
```suggestion
        stage('OpenSearch Debian APT promotion') {
```

gaiksaya commented on jenkins/promotion/release-promotion.jenkinsfile:
```diff
@@ -0,0 +1,433 @@
+                build job: 'distribution-promote-repos', wait: true, parameters: [choice(name: 'DISTRIBUTION_JOB_NAME', value: 'distribution-build-opensearch'), 
+                                                                                    choice(name: 'DISTRIBUTION_REPO_TYPE', value: 'apt'),
+                                                                                    string(name: 'DISTRIBUTION_BUILD_NUMBER', value: params.OPENSEARCH_RC),
+                                                                                    string(name: 'INPUT_MANIFEST', value: "${params.RELEASE_VERSION}/opensearch-${params.RELEASE_VERSION}.yml"),
```
Should we take the input manifest as paramter instead of dynamically forming the string? Maybe in next iteration? As of now cannot think of scenario we would name input manifest something else.

prudhvigodithi replied:
Sure, we can keep an eye on this and update in next iteration, just want to avoid a manual input for a release manager.

gaiksaya commented on jenkins/promotion/release-promotion.jenkinsfile:
```diff
@@ -0,0 +1,433 @@
+                                                                                    string(name: 'DISTRIBUTION_BUILD_NUMBER', value: params.OPENSEARCH_RC),
+                                                                                    string(name: 'INPUT_MANIFEST', value: "${params.RELEASE_VERSION}/opensearch-${params.RELEASE_VERSION}.yml"),
+                                                                                ]
+                echo 'Job distribution-build-opensearch completed!'
```
```suggestion
                echo 'Job distribution-promote-repos completed successfully!'
```
or 
```
echo 'OpenSearch <version> Debian/ APT repo released successfully'

gaiksaya commented on jenkins/promotion/release-promotion.jenkinsfile:
```diff
@@ -0,0 +1,433 @@
+                                                                                    string(name: 'DISTRIBUTION_BUILD_NUMBER', value: params.OPENSEARCH_RC),
+                                                                                    string(name: 'INPUT_MANIFEST', value: "${params.RELEASE_VERSION}/opensearch-${params.RELEASE_VERSION}.yml"),
+                                                                                ]
+                echo 'Job distribution-promote-repos for OpenSearch yum completed!'
```
```suggestion
                echo 'Job distribution-promote-repos for OpenSearch yum completed! OpenSearch <version> yum repo released successfully'
```

gaiksaya commented on jenkins/promotion/release-promotion.jenkinsfile:
```diff
@@ -0,0 +1,433 @@
+                echo 'Job distribution-promote-repos for OpenSearch yum completed!'
+            }
+        }
+        stage('OpenSearch Dashboards Apt promotion') {
```
```suggestion
        stage('OpenSearch Dashboards Debian APT promotion') {
```

gaiksaya commented on jenkins/promotion/release-promotion.jenkinsfile:
```diff
@@ -0,0 +1,433 @@
+                echo 'Job distribution-promote-artifacts for opensearch windows completed!'
+            }
+        }
+        stage('OpenSearch Linux deb arm64') {
```
```suggestion
        stage('OpenSearch Linux Debian arm64 artifacts') {
```

gaiksaya commented on tests/jenkins/TestOpenSearchReleasePromotionTest.groovy:
```diff
@@ -0,0 +1,45 @@
+    void shouldExecuteWithoutErrors() {
+        super.testPipeline('jenkins/promotion/release-promotion.jenkinsfile',
+                'tests/jenkins/jenkinsjob-regression-files/promotion/release-promotion.jenkinsfile')
+    }
```
Can we add some assertions to what all stages should be present at this job? Since this is a single point of execution I believe strong assertions are good to maintain. Any small change should break the tests indicating release can be blocked because of this change.

prudhvigodithi replied:
I have added multiple assertions for each stage and assertions with parameters as well, so that child promotion jobs are not triggered with wrong values when updating the jenkinsfile.

gaiksaya reviewed (commented):
Added some minor nits with naming. I would suggest to replace all echo statements with actually artifact promotion statments. For example, as an external user I don't know what promote-repos jobs does. So echo statement at the end of execution can help them understand, `yum repo for opensearch-2.12.0 is release successfully`.

prudhvigodithi commented:
Hey @gaiksaya I have updated all your NIT's in my latest commit, coming back with your suggestion to update the echo statement as `yum repo for opensearch-2.12.0 is release successfully`, the idea here is we are running the release promotion as part of the [pre-release activities](https://github.com/opensearch-project/opensearch-build/blob/main/RELEASE_PROCESS_OPENSEARCH.md#pre-release) and should be triggered by the [release-manager](https://github.com/opensearch-project/opensearch-build/blob/main/RELEASE_PROCESS_OPENSEARCH.md#release-manager), so the right way to know is if the child promotion job is successfully completed for not and not to ensure the distribution is released successfully, I have added updated the echo statements as example `Promotion successful for Debian Apt OpenSearch!`, let me know WDYT.
Thanks
